File: leetcode/198. House Robber/main.go
// Solved by @kitanoyoru

package main

func rob(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	p1, p2 := 0, 0

	for _, num := range nums {
		temp := p1
		if p2+num > p1 {
			p1 = p2 + num
		}
		p2 = temp
	}

	return p1
}

File: leetcode/198. House Robber/main.cpp
// Solved by @kitanoyoru
// https://leetcode.com/problems/house-robber/

class Solution {
public:
  int rob(vector<int>& nums) {
    const int n = nums.size();
    if (n == 0) return 0;

    int prev1 = 0, prev2 = 0;
    
    // prev2, prev1, num
    for (const int num : nums) {
      const int temp = prev1;
      prev1 = (prev2 + num > prev1) ? (prev2 + num) : prev1;
      prev2 = temp; 
    }

    return prev1;
  }
};
